### PR TITLE
Add tensor parallel sharding APIs for torch export

### PR DESCRIPTION
**Pull request by feikou**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111236

Add libraries to apply tensor parallel transformation to an exported program.

Differential Revision: [D50214796](https://our.internmc.facebook.com/intern/diff/D50214796/)

### CODE REVIEW

File: torch/distributed/_tensor/experimental/tensor_parallel.py [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+    input_specs_list: List[DTensorSpec] = []
+    for input_arg in node.all_input_nodes:
+        if input_arg in placement_strategies:
+            input_specs_list.append(placement_strategies[input_arg].output_spec)
+        else:
+            raise ValueError(f"{input_arg} does not have output_spec populated.")
+    return tuple(input_specs_list)
+
+
+def _get_arg_schema(
```
Comment (wanchaol) [inline torch/distributed/_tensor/experimental/tensor_parallel.py:385]: nit: maybe call this `_get_op_schema` and directly construct OpSchema instead

**Review by wanchaol — COMMENTED** [review]
This looks pretty good! I left a few nits, the main thing we should figure out is that we probably should not assume the even/odd layer of Linear and apply col/row parallel there, let's try to see if we can allow some level of module annotation in the exported program for better UX.

Another possible approach that I recently discussed with @SherlockNoMad  is that we can possibly first replace the input arguments with DTensor after the module annotation, then "re-export" the graph by going through the DTensor's torch_dispatch logic. But I feel we can make that a follow up change.

File: test/distributed/_tensor/experimental/test_tensor_parallel.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            self.mlps.append(
+                torch.nn.Sequential(
+                    torch.nn.Linear(6, 8),
+                    torch.nn.ReLU(),
+                    torch.nn.Linear(8, 6),
+                )
+            )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = torch.chunk(x, 2, dim=1)[0]
```
[the remark this answers is not in the document]
Comment (feikou) [reply test/distributed/_tensor/experimental/test_tensor_parallel.py:30]: This was to test special handling of getitem op in mark_sharding.

File: test/distributed/_tensor/experimental/test_tensor_parallel.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+class TensorParallelTest(DTensorTestBase):
+    def setUp(self) -> None:
+        super().setUp()
+
+    @with_comms
+    def test_tensor_parallel_transformation_e2e(self):
+        torch.manual_seed(0)
+        model = MLPListModule(2).to(device=self.device_type)
+        inputs = (torch.randn((10, 12)).to(device=self.device_type),)
+        with torch.inference_mode():
```
[the remark this answers is not in the document]
Comment (feikou) [reply test/distributed/_tensor/experimental/test_tensor_parallel.py:45]: Currently this is pretty commonly used in model processing and inference since gradient calculation is not needed. I'm curious why some aten ops may not decompose under inference mode. Is this something WIP or by design?

[the remark this answers is not in the document]
Comment (wanchaol) [reply test/distributed/_tensor/experimental/test_tensor_parallel.py:45]: Something it's a bit weird under torch.inference_mode(), for example, if you capture ops inside inference_mode, it would give you either `aten.linear.default` or `aten.matmul.default`, instead of giving you `aten.mm/addmm.default` instead. That is probably ok if torch.export is not run under `inference_mode`

[the remark this answers is not in the document]
Comment (feikou) [reply test/distributed/_tensor/experimental/test_tensor_parallel.py:45]: Interesting, that's good to know.

File: torch/distributed/_tensor/experimental/tensor_parallel.py [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+                    )
+                placeholder_idx += 1
+            else:
+                placement_strategies[node] = _create_placement_strategy(
+                    node,
+                    mesh,
+                    placements=(Replicate(),),
+                )
+    # If there are odd number of linear layers, no need to shard the params of the last layer.
+    if len(linear_node_names) % 2 == 1:
```
[the remark this answers is not in the document]
Comment (feikou) [reply torch/distributed/_tensor/experimental/tensor_parallel.py:122]: Sounds good. Will follow up to make tagging more generic.

File: torch/distributed/_tensor/experimental/tensor_parallel.py [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+                    op=op_overload,
+                    args_schema=_get_arg_schema(node, placement_strategies),
+                    kwargs_schema={},
+                )
+
+                # get DTensor specs for inputs and outputs
+                output_sharding = DTensor._propagator.propagate_op_sharding(
+                    op_schema,
+                )
+                # The stride from output_sharding may be incorrect for addmm (https://fburl.com/code/kb1ko758),
```
[the remark this answers is not in the document]
Comment (feikou) [reply torch/distributed/_tensor/experimental/tensor_parallel.py:175]: Yes, tried that the PR worked well for this case!

File: torch/distributed/_tensor/experimental/tensor_parallel.py [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+
+
+aten = torch.ops.aten
+
+
+def tensor_parallel_transformation(
+    exported_program: ExportedProgram,
+    rank: int,
+    world_size: int,
+) -> ExportedProgram:
```
Comment (wanchaol) [inline torch/distributed/_tensor/experimental/tensor_parallel.py:37]: Let's add a docstr to:
* explicitly mention that this is a graph transformation pass for torch.export to transform a single device graph to a tensor parallel graph
* add a warning here that this API is experimental and subject to change so that people don't abuse this API

File: torch/distributed/_tensor/experimental/tensor_parallel.py [hunk 7]
```diff
@@ -0,0 +1,426 @@
+import copy
```
Comment (wanchaol) [inline torch/distributed/_tensor/experimental/tensor_parallel.py:1]: nit: let's also rename this file to sth like `tp_transform.py` or `tensor_parallel_transform.py` to make this more explicit that this is a graph transformation pass.

File: torch/distributed/_tensor/experimental/tensor_parallel.py [hunk 8]
```diff
@@ -0,0 +1,426 @@
...
+
+        placement_strategies = mark_sharding(gm, self.graph_signiture, self.mesh)
+        partitioner(gm)
+        shard_state_dict(
+            self.state_dict, placement_strategies, self.graph_signiture, self.mesh
+        )
+        return PassResult(gm, True)
+
+
+def mark_tensor_parallel_shardings(
```
Comment (wanchaol) [inline torch/distributed/_tensor/experimental/tensor_parallel.py:73]: maybe we should made all those APIs be private (i.e. by prepanding `_`) as they are all exclusive to this file only. let's only leave a public `tensor_parallel_transformation` API and make the rest to be private

**Review by wanchaol — APPROVED** [review]
lgtm, great work! Left a few comments to add some docstr and make some APIs be private.
